scripts/changesummary.py

```diff
@@ -0,0 +1,540 @@
+            # to make it easier for maintainers to see only the relevant changes
+            # when dicovery artifacts are updated rather than the changes due to
+            # random key ordering.
+            self._rewrite_as_sorted_json(file_path)
+
+            with open(file_path, "r+") as f:
+                # Now load the json file into a pandas dataframe as a flat table
+                dataframe_doc = pd.json_normalize(json.load(f))
+        return dataframe_doc
+
+    def _get_discovery_differences(self, filename):
+        """Returns a pandas dataframe which contains the differences with the
+        current and new discovery artifact directories, corresponding to the
+        file name provided.
+
+        args:
+            filename (str): The name of the discovery artifact to parse.
+        """
+        # The paths of the 2 discovery artifacts to compare
+        current_artifact_path = os.path.join(self._current_artifacts_dir, filename)
+        new_artifact_path = os.path.join(self._new_artifacts_dir, filename)
+
+        # Use a helper functions to load the discovery artifacts into pandas
+        # dataframes
+        current_doc = self._load_json_to_dataframe(current_artifact_path)
+        new_doc = self._load_json_to_dataframe(new_artifact_path)
+
+        # Concatenate the 2 dataframes, transpose them, and create
+        # a new dataframe called combined_docs with columns
+        # `Keys`, `CurrentValue`, `NewValue`.
+        combined_docs = (
+            pd.concat([current_doc, new_doc], keys=["CurrentValue", "NewValue"])
+            .transpose()
+            .rename_axis("Key")
+            .reset_index()
+        )
+
+        # This line is for cleanup purposes. As a result of the previous code
+        # we have a multi-index which is not needed so we need to drop a level.
+        combined_docs.columns = combined_docs.columns.droplevel(1)
+
+        # When discovery documents are added, the column `CurrentValue` will
+        # not exist. In that case, we'll just populate with `np.nan`.
+        if "CurrentValue" not in combined_docs.columns:
+            combined_docs["CurrentValue"] = np.nan
+
+        # When discovery documents are deleted, the column `NewValue` will
+        # not exist. In that case, we'll just populate with `np.nan`.
+        if "NewValue" not in combined_docs.columns:
+            combined_docs["NewValue"] = np.nan
+
+        # Split the Key into 2 columns for `Parent` and `Child` in order
+        # to group keys with the same parents together to summarize the changes
+        # by parent.
+        parent_child_df = combined_docs["Key"].str.rsplit(".", 1, expand=True)
+        # Rename the columns and join them with the combined_docs dataframe.
+        # If we only have a `Parent` column, it means that the Key doesn't have
+        # any children.
+        if len(parent_child_df.columns) == 1:
+            parent_child_df.columns = ["Parent"]
```

Comment:
I haven't seen columns use a setter before. This renames the column(s)?

Reply:
Correct. I see the note below in the [docs](https://pandas.pydata.org/pandas-docs/stable/user_guide/basics.html#attributes-and-underlying-data) with an example.

>Note, these attributes can be safely assigned to!
```
In [7]: df[:2]
Out[7]: 
                   A         B         C
2000-01-01 -0.173215  0.119209 -1.044236
2000-01-02 -0.861849 -2.104569 -0.494929

In [8]: df.columns = [x.lower() for x in df.columns]

In [9]: df
Out[9]: 
                   a         b         c
2000-01-01 -0.173215  0.119209 -1.044236
```